pkg/handlers/generic/mutation/controlplanevirtualip/inject_test.go

```diff
@@ -268,3 +275,247 @@ spec:
         - name: port
           value: "{{ .Port }}"
 `
+
+func Test_deleteFiles(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name          string
+		obj           *controlplanev1.KubeadmControlPlaneTemplate
+		filesToDelete []string
+		expectedFiles []bootstrapv1.File
+	}{
+		{
+			name: "should delete files from the template",
+			obj: &controlplanev1.KubeadmControlPlaneTemplate{
+				Spec: controlplanev1.KubeadmControlPlaneTemplateSpec{
+					Template: controlplanev1.KubeadmControlPlaneTemplateResource{
+						Spec: controlplanev1.KubeadmControlPlaneTemplateResourceSpec{
+							KubeadmConfigSpec: bootstrapv1.KubeadmConfigSpec{
+								Files: []bootstrapv1.File{
+									{
+										Path: "file-1",
+									},
+									{
+										Path: "file-2",
+									},
+									{
+										Path: "file-3",
+									},
+									{
+										Path: "file-4",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			filesToDelete: []string{"file-2", "file-3", "file-5"},
+			expectedFiles: []bootstrapv1.File{
+				{
+					Path: "file-1",
+				},
+				{
+					Path: "file-4",
+				},
+			},
+		},
+		{
+			name: "should keep all files",
+			obj: &controlplanev1.KubeadmControlPlaneTemplate{
+				Spec: controlplanev1.KubeadmControlPlaneTemplateSpec{
+					Template: controlplanev1.KubeadmControlPlaneTemplateResource{
+						Spec: controlplanev1.KubeadmControlPlaneTemplateResourceSpec{
+							KubeadmConfigSpec: bootstrapv1.KubeadmConfigSpec{
+								Files: []bootstrapv1.File{
+									{
+										Path: "file-1",
+									},
+									{
+										Path: "file-2",
+									},
+									{
+										Path: "file-3",
+									},
+									{
+										Path: "file-4",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			filesToDelete: []string{"file-5"},
+			expectedFiles: []bootstrapv1.File{
+				{
+					Path: "file-1",
+				},
+				{
+					Path: "file-2",
+				},
+				{
+					Path: "file-3",
+				},
+				{
+					Path: "file-4",
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			deleteFiles(tt.obj, tt.filesToDelete...)
+			assert.Equal(t, tt.expectedFiles, tt.obj.Spec.Template.Spec.KubeadmConfigSpec.Files)
```

Comment:
Consider using ElementsMatch so test doesn't rely on order
```go
assert.ElementsMatch(
				t,
				tt.expectedFiles,
				tt.obj.Spec.Template.Spec.KubeadmConfigSpec.Files,
			)
```

Reply:
We do want to preserve the order, please see https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/1062#discussion_r1989907485